pkg/streamingpromql/optimize/plan/querysplitting/node.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querysplitting
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/planning"
+	"github.com/grafana/mimir/pkg/streamingpromql/planning/core"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+func init() {
+	planning.RegisterNodeFactory(func() planning.Node {
+		return &SplittableFunctionCall{SplittableFunctionCallDetails: &SplittableFunctionCallDetails{}}
+	})
+}
+
+// SplittableFunctionCall wraps a range vector function call to split its computation into
+// fixed-interval blocks for intermediate result caching.
+type SplittableFunctionCall struct {
+	*SplittableFunctionCallDetails
+	Inner planning.Node
+}
+
+func (s *SplittableFunctionCall) Details() proto.Message {
+	return s.SplittableFunctionCallDetails
+}
+
+func (s *SplittableFunctionCall) NodeType() planning.NodeType {
+	return planning.NODE_TYPE_SPLIT_RANGE_VECTOR
+}
+
+func (s *SplittableFunctionCall) Children() []planning.Node {
+	return []planning.Node{s.Inner}
+}
+
+func (s *SplittableFunctionCall) SetChildren(children []planning.Node) error {
+	if len(children) != 1 {
+		return fmt.Errorf("node of type SplittableFunctionCall supports 1 child, but got %d", len(children))
+	}
+
+	s.Inner = children[0]
+	return nil
+}
+
+func (s *SplittableFunctionCall) EquivalentTo(other planning.Node) bool {
+	otherSplit, ok := other.(*SplittableFunctionCall)
+	return ok && s.Inner.EquivalentTo(otherSplit.Inner)
+}
+
+func (s *SplittableFunctionCall) Describe() string {
+	return "split into cacheable blocks"
+}
+
+func (s *SplittableFunctionCall) ChildrenLabels() []string {
+	return []string{""}
+}
+
+func (s *SplittableFunctionCall) ChildrenTimeRange(parentTimeRange types.QueryTimeRange) types.QueryTimeRange {
+	return parentTimeRange
+}
+
+func (s *SplittableFunctionCall) ResultType() (parser.ValueType, error) {
+	return s.Inner.ResultType()
+}
+
+func (s *SplittableFunctionCall) QueriedTimeRange(queryTimeRange types.QueryTimeRange, lookbackDelta time.Duration) planning.QueriedTimeRange {
+	return s.Inner.QueriedTimeRange(queryTimeRange, lookbackDelta)
+}
+
+func (s *SplittableFunctionCall) ExpressionPosition() posrange.PositionRange {
+	return s.Inner.ExpressionPosition()
+}
+
+func (s *SplittableFunctionCall) MinimumRequiredPlanVersion() planning.QueryPlanVersion {
+	return s.Inner.MinimumRequiredPlanVersion()
```

Comment:
We'll need to introduce a new query plan version, given this is a new kind of node.